### PR TITLE
Claim GitHub App installations on first login if they exist

### LINKED ISSUE #2521: Support linking installed organizations to a project (Marketplace flow)
Depends on: https://github.com/stacklok/infra/issues/831

We need to support the use case where a user installs the Minder GitHub App on their GitHub organization without having created a Minder account (or simply not being logged in).

We should give Minder user the ability to "link" (open to better terminology) an GitHub organization to their project, if they are the owner of the GitHub organization and if the Minder app is installed on the organization.

When a user logs in, we can get their GitHub user ID from their JWT. We can then query all "unlinked" organization to see if the user owns any of them and present them with the option to add an organization to their project.

See [Linking an enrolled organization to your project](https://docs.google.com/document/d/1SvXgkaB8ffgD2VH9pQq8rabgV-8ow3T9oBRtgZFPGVY/edit#heading=h.lcq7pgxnvu1r) in the design doc.

### LINKED ISSUE #2685: Claim pending Providers when a new user is created.
If a user has logged in to Minder before they do a GitHub-side app installation flow, then we can match the app installation from the end of the OAuth flow with their account.  If there are no matching accounts at that time, we store the pending installation (completed on the GitHub side, but not matched in Minder) in the `provider_github_app_installations` table.

When a user logs in, we should check the `provider_github_app_installations` table for any pending installations -- if these exist, we should create projects & providers for each, rather than creating the default `$username` project.

### PR DESCRIPTION
**Pull request by evankanderson**
# Summary

Currently, if a user enrolls a GitHub App before they log in to Minder and create an account, we store the installation information in provider_github_app_installations, but there's no way for users to get at it.  Worse, the presence of that row will prevent subsequent re-enrollment of that repo if we miss the app deletion webhook.

This PR changes the CreateUser flow for users with a pending GitHub app installation to create one project per pending installation when they log in, rather than creating the default OAuth provider.

Fixes #2685

## Change Type

***Mark the type of change your PR introduces:***

- [ ] Bug fix (resolves an issue without affecting existing features)
- [x] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [x] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

Manual testing of the new flow, along with unit testing of existing flows / method refactors.

Will be adding tests for the new login flows shortly, but wanted to get this out for review early.

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [x] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [ ] Included tests that validate the fix or feature.
- [x] Checked that related changes are merged.

### CODE REVIEW

File: internal/providers/service.go [hunk 1]
```diff
@@ -235,66 +244,56 @@ func (p *providerService) CreateGitHubAppProvider(
 	})
 }
 
-// CreateGitHubAppWithoutInvitation creates an GitHub App installation that doesn't belong to a project yet
+// CreateGitHubAppWithoutInvitation either creates a new project for the selected app, or stores
```
Comment (eleftherias) [inline internal/providers/service.go:247]: nit:
```suggestion
// CreateGitHubAppWithoutInvitation either creates a new project for the selected app installation, or stores
```

**Review by eleftherias — COMMENTED** [review]
Works really well! I left a few small comments and it looks like the lint job is failing.

File: internal/controlplane/server.go [hunk 2]
```diff
@@ -177,6 +179,7 @@ func NewServer(
 		repos:               github.NewRepositoryService(whManager, store, evt),
 		marketplace:         marketplace,
 		providerStore:       providerStore,
+		ghClient:            &ghprov.ClientServiceImplementation{},
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply internal/controlplane/server.go:182]: @dmjb -- do you have opinions on where I move this?

It felt a little icky to have it here, but I couldn't figure out a better location.

File: internal/providers/service.go [hunk 3]
```diff
@@ -235,66 +244,56 @@ func (p *providerService) CreateGitHubAppProvider(
...
 		})
 		if err != nil {
 			return nil, fmt.Errorf("error saving installation ID: %w", err)
 		}
-		return &gitHubAppInstallation, nil
+		return nil, nil
 	}
 
-	provider, err := createGitHubApp(ctx, qtx, project.ID, installationOwner, installationID, nil, sql.NullString{})
+	zerolog.Ctx(ctx).Info().Str("project", project.ID.String()).Int64("org", installationOwner.GetID()).
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply internal/providers/service.go]: :+1:

File: internal/controlplane/handlers_oauth.go [hunk 4]
```diff
@@ -381,6 +377,33 @@ func (s *Server) exchangeCodeForToken(ctx context.Context, providerName string,
...
+	qtx := s.store.GetQuerierWithTransaction(tx)
+
+	_, err = s.providers.CreateGitHubAppWithoutInvitation(ctx, qtx, *userID, installationID)
+	if err != nil {
+		return fmt.Errorf("error handling app without minder state: %w", err)
+	}
+
+	if err = s.store.Commit(tx); err != nil {
+		return fmt.Errorf("error committing transaction: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (evankanderson) [reply internal/controlplane/handlers_oauth.go]: Thanks!  I feel like I'd had trouble with this and mocks in the past, but it seems to be working fine in the tests now.

**Review by eleftherias — APPROVED** [review]

**Pull request merged**